### PR TITLE
[2/6] feat(buildings): add the CreateBuildings batch RPC

### PR DESCRIPTION
**Pull request by flesher**
Reviewable diff: +373/-0 across 5 files (excludes generated, test, and story files).

## Summary

Adds `CreateBuildings`, which creates several buildings in one site in a single all-or-nothing transaction. This is what lets the operator UI generate a name series (prefix + counter) and create 40 buildings in one action instead of 40 round trips.

Backend + generated types only. The form that calls it arrives in #864, wired to a host in #860.

### Stack — 6 PRs, merge in order

| # | PR | What it adds |
|---|----|--------------|
| 1 | #855 | `AssignDevicesToRack` also writes rack slot placement |
| **2** | **#856 ← this PR** | `CreateBuildings` batch RPC |
| 3 | #862 | `CreateRacks` batch RPC |
| 4 | #863 | Rack create modal gains a Multiple mode |
| 5 | #864 | Building surfaces: create racks and buildings inline |
| 6 | #860 | Site surfaces: create buildings inline |

Together the series lets an operator build out the physical hierarchy without leaving the modal they are in — buildings from inside a site, racks from inside a building — and create them in batches from a generated name series. PRs 1–3 are the backend RPCs; 4–6 are the UI that calls them.

## How it works

The request carries a required `site_id` and 1–500 `NewBuilding` rows in the order the UI previewed them. The UI sends finished names rather than generation parameters, so the preview it shows and the rows created cannot diverge.

Names are checked twice — within the batch, and against the buildings already at `site_id`. A collision rejects the whole request and creates nothing, returning one `PerBuildingCreateError` per offending row with its zero-based index, the rejected name, and a reason (`DUPLICATE_NAME_IN_BATCH` or `DUPLICATE_NAME_AT_SITE`) so the UI can mark exactly those rows instead of failing the form opaquely.

Either every building is created or none is, matching `CreateBuilding`'s existing stance on its seeds (#559): a partial batch leaves the operator reconciling which half of the list exists.

```mermaid
flowchart TD
  A["CreateBuildings(site_id, [NewBuilding])"] --> B["validate: names unique in batch"]
  B --> C["validate: names unused at site"]
  C -->|collision| D["errors[], nothing created"]
  C -->|ok| E["insert all rows in one tx"]
  E --> F["buildings[] in request order"]
```

## Areas of the code involved

| Area | What changed | Why it matters |
|---|---|---|
| `proto/buildings/v1/buildings.proto` | `CreateBuildings`, `NewBuilding`, per-row error types | The contract, including the 500-row cap and field bounds |
| `server/internal/domain/buildings/service.go` | Batch validation + transactional insert | The load-bearing logic |
| `server/internal/domain/buildings/models/models.go` | `NewBuilding` domain model | |
| `server/internal/handlers/buildings/{handler,translate}.go` | RPC wiring and proto↔domain mapping | |
| `server/generated/**`, `client/**/generated/**` | regen — skip | |

A second commit in this PR adds `aisles` / `racks_per_aisle` to `NewBuilding`, matching what single create already accepts. Additive — a request omitting them behaves as before.

## Key technical decisions

- **`site_id` is required here**, unlike `CreateBuilding` — the only entry point is bulk create from inside a site, and the UI has no orphan-building create path.
- **All-or-nothing, with per-row errors**, rather than partial success — the operator gets a form they can fix, not a half-applied list to reconcile.
- **No rack or device seeding**, unlike single create — seeding N buildings at once has no sensible way to say which rack belongs to which building.
- **500-row cap** as a typo guard, so a runaway counter can't open a 100k-row transaction.

## Testing & validation

- `go test ./internal/domain/buildings/... ./internal/handlers/buildings/...` passes, covering in-batch duplicates, existing-name collisions, rollback, and order preservation.
- `go build ./...` and `tsc --noEmit` clean.
- Not covered: no end-to-end run — nothing calls this RPC until #864/#860.

### CODE REVIEW

File: proto/buildings/v1/buildings.proto [hunk 1]
```diff
@@ -44,6 +44,28 @@ service BuildingService {
...
+    // the buildings already at site_id. A violation rejects the whole
+    // request with one `errors` entry per offending index and creates
+    // nothing, so the UI can mark the offending rows.
+    //
+    // Unlike CreateBuilding, site_id is REQUIRED — the only entry point
+    // is bulk-create from inside a site, and the plan is explicit that
+    // the UI has no orphan-building create paths. There is deliberately
+    // no rack/device seeding here: seeding N buildings at once has no
+    // sensible way to say which rack belongs to which building.
+    rpc CreateBuildings(CreateBuildingsRequest) returns (CreateBuildingsResponse);
```
[the remark this answers is not in the document]
Comment (flesher) [reply proto/buildings/v1/buildings.proto:52]: 🤖 Correct, and it was already red — thanks. Confirmed by running it before fixing:

```
--- FAIL: TestRPCContract_EveryRegisteredProcedureIsClassified
    Should be empty, but was [/buildings.v1.BuildingService/CreateBuildings]
```

Added at `site:manage` in `81cbbf07`, with a note that unlike `CreateBuilding` there is no `rack:manage` escalation to gate here (bulk create takes no rack/device seed).

File: server/internal/handlers/buildings/handler.go [hunk 2]
```diff
@@ -100,6 +100,28 @@ func (h *Handler) CreateBuilding(ctx context.Context, req *connect.Request[pb.Cr
 	}), nil
 }
 
+func (h *Handler) CreateBuildings(ctx context.Context, req *connect.Request[pb.CreateBuildingsRequest]) (*connect.Response[pb.CreateBuildingsResponse], error) {
+	info, err := middleware.RequirePermission(ctx, authz.PermSiteManage, authz.ResourceContext{})
```
[the remark this answers is not in the document]
Comment (flesher) [reply server/internal/handlers/buildings/handler.go]: 🤖 Valid on both halves. Traced it through `EffectivePermissions.Has`: with a nil `SiteID` it consults `orgScope` only, so a site-scoped `site:manage` operator was denied at their own site; and it never reaches the `bySite` lookup, so an assignment narrowing an org-wide grant at the target site was skipped. `site_id` is required (`gt = 0`) for bulk create, so the scope is unambiguous.

Fixed in `81cbbf07`, with tests for both directions: a site-scoped manager is admitted at site 42 and denied at site 99.

One note — the pre-existing single `CreateBuilding` (handler.go:68) passes an empty `ResourceContext` too, so it has the same gap. I left it alone to keep this PR scoped rather than change auth behavior on an established RPC in a bulk-create PR. Happy to file it separately.

File: server/internal/domain/buildings/service.go [hunk 3]
```diff
@@ -287,6 +288,156 @@ func (s *Service) CreateBuilding(ctx context.Context, params models.CreateParams
...
+			row, err := s.store.CreateBuilding(txCtx, models.CreateParams{
+				OrgID:       params.OrgID,
+				SiteID:      &siteID,
+				Name:        names[i],
+				Description: b.Description,
+				PowerKw:     b.PowerKw,
+				OverheadKw:  b.OverheadKw,
+
+				Aisles:        b.Aisles,
+				RacksPerAisle: b.RacksPerAisle,
```
[the remark this answers is not in the document]
Comment (flesher) [reply server/internal/domain/buildings/service.go:390]: 🤖 Fixed in `81cbbf07` — each row now goes through `validateLayoutBounds` before the tx opens, naming the offending index.

Worth recording that the exposure is narrower than it reads: `NewBuilding.aisles` and `racks_per_aisle` already carry `(buf.validate.field).int32 = {gte: 0, lte: 100}`, the same cap as `layoutDimensionMax`, so no RPC caller could exceed it. This is defense in depth for a direct service caller. Added as such since it costs a loop and keeps bulk and single consistent, which was your point.

File: server/internal/domain/buildings/service.go [hunk 4]
```diff
@@ -287,6 +288,156 @@ func (s *Service) CreateBuilding(ctx context.Context, params models.CreateParams
...
+	names := make([]string, len(params.Buildings))
+	for i, b := range params.Buildings {
+		names[i] = strings.TrimSpace(b.Name)
+		if names[i] == "" {
+			return nil, nil, fleeterror.NewInvalidArgumentErrorf("buildings[%d].name is required", i)
+		}
+	}
+
+	if dupes := duplicateNamesInBatch(names); len(dupes) > 0 {
+		return nil, dupes, nil
```
[the remark this answers is not in the document]
Comment (flesher) [reply server/internal/domain/buildings/service.go]: 🤖 Fixed in `81cbbf07`. The site check now always runs and the two sets are merged into one entry per row, ordered by index. For your exact example — `["B-001", "B-001"]` with `B-001` live at the site — the response is now index 0 `DUPLICATE_NAME_AT_SITE` and index 1 `DUPLICATE_NAME_IN_BATCH`, so one submit shows both. A row that is both reports AT_SITE, since renaming is required either way and that is the reason editing other rows cannot resolve.

Two consequences worth flagging rather than burying: batch duplicates no longer short-circuit, so an already-invalid request now costs one tx and a site-row lock (the happy path is unchanged); and such a request against a missing site now returns NotFound instead of the duplicate list, which I think is the better ordering. Two tests asserted the old "before opening tx" contract and were updated; `TestCreateBuildings_reportsBothUniquenessSourcesInOneResponse` covers the new one.

File: server/internal/domain/buildings/service.go [hunk 5]
```diff
@@ -287,6 +288,156 @@ func (s *Service) CreateBuilding(ctx context.Context, params models.CreateParams
...
+				Name:        names[i],
+				Description: b.Description,
+				PowerKw:     b.PowerKw,
+				OverheadKw:  b.OverheadKw,
+
+				Aisles:        b.Aisles,
+				RacksPerAisle: b.RacksPerAisle,
+			})
+			if err != nil {
+				return nil, err
```
[the remark this answers is not in the document]
Comment (flesher) [reply server/internal/domain/buildings/service.go:406]: 🤖 Confirmed reachable and fixed in `81cbbf07`. Checked the premise rather than assuming: `WithTransaction` passes nil `TxOptions`, so these run at READ COMMITTED, and `UpdateBuilding` takes `LockBuildingForWrite` on the building row while bulk create locks the site row — different rows, no mutual exclusion. So a concurrent rename can commit between the `ListBuildings` read and our insert with no serialization failure to retry on.

The store's `AlreadyExists` now becomes that row's `DUPLICATE_NAME_AT_SITE` (matching on `GRPCCode`, following the `fleetimport/importer.go:30` idiom), so rollback is preserved and the UI can still mark the row. Any other error still propagates. Test: `TestCreateBuildings_insertRaceBecomesPerRowRejection`.

File: server/internal/domain/buildings/service.go [hunk 6]
```diff
@@ -287,6 +290,213 @@ func (s *Service) CreateBuilding(ctx context.Context, params models.CreateParams
...
+		// below can't race another create at the same site. NotFound surfaces
+		// directly. Same guard createBuildingInTx applies for a single create.
+		if err := s.siteStore.LockSiteForWrite(txCtx, params.OrgID, params.SiteID); err != nil {
+			return nil, err
+		}
+
+		existing, err := s.store.ListBuildings(txCtx, models.ListFilter{
+			OrgID:   params.OrgID,
+			SiteIDs: []int64{params.SiteID},
+		})
```
[the remark this answers is not in the document]
Comment (flesher) [reply server/internal/domain/buildings/service.go]: 🤖 Valid — fixed in `978519c0`. The preflight now calls a names-only `ListBuildingNamesBySite` (scoped to the target site, `deleted_at IS NULL`) instead of `ListBuildings`, so it no longer drags in `ListBuildingsByOrg`'s two org-wide `LEFT JOIN` aggregations (rack_count + device_count over `device_set_rack`/`device_set`/`device_set_membership`/`device`) while the site row is write-locked.

Semantics are unchanged: `site_id` is required (`gt = 0`) for bulk create, so the site-scoped query matches the partial unique index on `(site_id, name)` exactly — no unassigned-building edge case. Covered by the existing `sqlstores` integration test against the real DB.

File: server/internal/handlers/buildings/handler_test.go [hunk 7]
```diff
@@ -512,3 +512,120 @@ func TestHandler_AssignRacksToBuilding_happy(t *testing.T) {
...
 
 func ptrInt32t(v int32) *int32 { return &v }
+
+func TestHandler_CreateBuildings_happy(t *testing.T) {
+	t.Parallel()
+	h := newTestHandler(t)
+
+	h.siteStore.EXPECT().LockSiteForWrite(gomock.Any(), int64(7), int64(42)).Return(nil)
+	h.buildingStore.EXPECT().ListBuildings(gomock.Any(), gomock.AssignableToTypeOf(models.ListFilter{})).
+		Return([]models.BuildingWithCounts{}, nil)
```
[the remark this answers is not in the document]
Comment (flesher) [reply server/internal/handlers/buildings/handler_test.go]: 🤖 Good catch — real regression, fixed in `255976fe`. The names-only preflight swapped the production path to `ListBuildingNamesBySite`, but these three handler tests (522/551/601) still stubbed `ListBuildings`; they now stub the new method. Confirmed `go test ./internal/handlers/buildings/...` is green.

File: proto/buildings/v1/buildings.proto [hunk 8]
```diff
@@ -254,6 +276,76 @@ message CreateBuildingResponse {
...
 
+// CreateBuildings
+
+// One building in a bulk-create batch. Carries only the fields the bulk
+// form can set: the operator names the buildings (generated from a prefix
+// and counter, or typed) and everything else takes the same server-side
+// zero default a plain CreateBuilding would. Layout dimensions and rack
+// defaults are deliberately absent — they are per-building concerns the
+// operator sets afterwards on the building itself, and baking one layout
+// into a batch of 40 would be wrong more often than right.
```
[the remark this answers is not in the document]
Comment (flesher) [reply proto/buildings/v1/buildings.proto]: 🤖 Fixed in `255976fe`. Updated the `NewBuilding` doc comment to say layout dimensions (aisles, racks_per_aisle) are carried per row — see the field comments — while rack defaults remain absent. Regenerated the pb comment in `buildings.pb.go` / `buildings_pb.ts` in lockstep.

File: proto/device_set/v1/device_set.proto [hunk 9]
```diff
@@ -90,6 +90,11 @@ service DeviceSetService {
     // device_selector accepts only the device_list variant. all_devices
     // is rejected with InvalidArgument because moving every paired
     // device into a single rack is never the intended operation.
+    //
+    // slot_assignments optionally carries each moved device's slot, so
+    // membership and placement land in one transaction. Prefer this over
+    // SaveRack for every edit — SaveRack replaces the rack's whole member
+    // set, so a stale client snapshot silently drops concurrent additions.
     rpc AssignDevicesToRack(AssignDevicesToRackRequest) returns (AssignDevicesToRackResponse);
```
[the remark this answers is not in the document]
Comment (flesher) [reply proto/device_set/v1/device_set.proto:98]: 🤖 Resolved by rebase. This PR was stacked on #855 (`[1/6]` rack slot placement) while its base was still `main`, so the diff showed #855's `AssignDevicesToRack` changes too. #855 is now merged; I rebased #856 onto the new `main`, so the diff is buildings-only and these rack changes are gone from here.

File: server/internal/domain/collection/service.go [hunk 10]
```diff
@@ -1277,6 +1449,19 @@ func (s *Service) AssignDevicesToRack(ctx context.Context, params AssignDevicesT
...
 			}
 			newlyAssigned = added
+
+			// Capacity guard: every rack member is expected to occupy a
+			// slot, so membership is bounded by rows×columns — the same
+			// invariant SaveRack enforces on its replace path. prior +
+			// newly-inserted is exact, since `added` counts only rows the
+			// ON CONFLICT DO NOTHING insert created.
+			if capacity := int64(targetRows) * int64(targetColumns); capacity > 0 {
+				if resulting := int64(targetPriorCount) + added; resulting > capacity {
```
[the remark this answers is not in the document]
Comment (flesher) [reply server/internal/domain/collection/service.go:1459]: 🤖 Now out of scope for this PR — after rebasing onto `main` (which merged #855), this `AssignDevicesToRack` code is no longer in #856's diff. The stale-member capacity concern is worth a look, but it lives in merged `main` now, so it belongs in a follow-up against #855's code rather than here. Resolving on this PR.

**Review by ankitgoswami — APPROVED** [review]
